Implement ImageBufferPipe for CoreAnimation so that we can update frames without using the main-thread

mattwoodrow opened the pull request:
#### f92c75b5eb3f5701b5a37283816170d8f861e829
<pre>
Implement ImageBufferPipe for CoreAnimation so that we can update frames without using the main-thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=244878">https://bugs.webkit.org/show_bug.cgi?id=244878</a>
&lt;rdar://problem/99902985&gt;

Creates a new API for pushing contents to a layer. The requirements here are that it works asynchronously (from a worker thread),
and that it works with UI process compositing (can&apos;t access any platform layers).

The existing APIs are insufficient, since setContentsToPlatformLayer won&apos;t work in the WebProcess if UI-side compositing is enabled,
and setContentsDisplayDelegate has main-thread callbacks for the delegate to provide new contents.

This adds a new API GraphicsLayerAsyncContentsDisplayDelegate, which get allocated on-demand by the GraphicsLayer implementation and
allows the consumer to push an ImageBuffer that will be copied into the GraphicsLayer asynchronously.

It adds an implementation of this new delegate type for GraphicsLayerCA, and a non-nicosia implementation of ImageBufferPipe which
uses it for frame transfer.

ImageBufferPipe (and the new delegate) both work by providing an ImageBuffer reference that needs to be copied if it succeeds, so that we
never end up copying in the caller and not using the frame.

Reviewed by Simon Fraser.

* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::usesPlatformLayer const):
(WebCore::CanvasRenderingContext::platformLayer):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
(WebCore::PlaceholderRenderingContext::usesPlatformLayer const):
(WebCore::PlaceholderRenderingContext::platformLayer):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.h:
* Source/WebCore/platform/graphics/ImageBufferPipe.cpp:
* Source/WebCore/platform/graphics/ImageBufferPipe.h:
(WebCore::ImageBufferPipe::usesPlatformLayer const):
(WebCore::ImageBufferPipe::platformLayer):
* Source/WebCore/platform/graphics/cocoa/ImageBufferPipeCA.mm: Added.
(WebCore::ImageBufferPipeSourceCA::ImageBufferPipeSourceCA):
(WebCore::ImageBufferPipeCA::ImageBufferPipeCA):
(WebCore::ImageBufferPipe::create):
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):

Canonical link: <a href="https://commits.webkit.org/257696@main">https://commits.webkit.org/257696@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/037640b9da7835b15254ccfbcc5f7daca944dfec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99717 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8893 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32796 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109079 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169312 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103720 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9535 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86182 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92177 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106987 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105486 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90656 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34111 "An unexpected error occured. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Pull request contains relevant changes; Skipped layout-tests; layout-tests (exception)") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22028 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2715 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23542 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2661 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45924 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5290 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8797 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43016 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4521 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
```diff
@@ -197,6 +197,8 @@ class GraphicsLayerCA : public GraphicsLayer, public PlatformCALayerClient {
 
     constexpr static CompositingCoordinatesOrientation defaultContentsOrientation = CompositingCoordinatesOrientation::TopDown;
 
+    WEBCORE_EXPORT RefPtr<GraphicsLayerAsyncContentsDisplayDelegate> getAsyncContentsDisplayDelegate() override;
```
This should be a "create" since it makes a new one every time.

smfr commented on Source/WebCore/platform/graphics/ca/cocoa/GraphicsLayerAsyncContentsDisplayDelegateCocoa.h:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2014 Apple Inc. All rights reserved.
```
2014?

mattwoodrow committed:
Implement ImageBufferPipe for CoreAnimation so that we can update frames without using the main-thread
https://bugs.webkit.org/show_bug.cgi?id=244878
<rdar://problem/99902985>

Creates a new API for pushing contents to a layer. The requirements here are that it works asynchronously (from a worker thread),
and that it works with UI process compositing (can't access any platform layers).

The existing APIs are insufficient, since setContentsToPlatformLayer won't work in the WebProcess if UI-side compositing is enabled,
and setContentsDisplayDelegate has main-thread callbacks for the delegate to provide new contents.

This adds a new API GraphicsLayerAsyncContentsDisplayDelegate, which get allocated on-demand by the GraphicsLayer implementation and
allows the consumer to push an ImageBuffer that will be copied into the GraphicsLayer asynchronously.

It adds an implementation of this new delegate type for GraphicsLayerCA, and a non-nicosia implementation of ImageBufferPipe which
uses it for frame transfer.

ImageBufferPipe (and the new delegate) both work by providing an ImageBuffer reference that needs to be copied if it succeeds, so that we
never end up copying in the caller and not using the frame.

Reviewed by Simon Fraser.

* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::usesPlatformLayer const):
(WebCore::CanvasRenderingContext::platformLayer):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.cpp:
(WebCore::PlaceholderRenderingContext::usesPlatformLayer const):
(WebCore::PlaceholderRenderingContext::platformLayer):
* Source/WebCore/html/canvas/PlaceholderRenderingContext.h:
* Source/WebCore/platform/graphics/ImageBufferPipe.cpp:
* Source/WebCore/platform/graphics/ImageBufferPipe.h:
(WebCore::ImageBufferPipe::usesPlatformLayer const):
(WebCore::ImageBufferPipe::platformLayer):
* Source/WebCore/platform/graphics/cocoa/ImageBufferPipeCA.mm: Added.
(WebCore::ImageBufferPipeSourceCA::ImageBufferPipeSourceCA):
(WebCore::ImageBufferPipeCA::ImageBufferPipeCA):
(WebCore::ImageBufferPipe::create):
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):

Canonical link: https://commits.webkit.org/257696@main
